src/SAML2/Extensions/GsspUserAttributesChunk.php

```diff
@@ -0,0 +1,63 @@
+<?php
+/**
+ * Copyright 2021 SURF B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\SamlBundle\SAML2\Extensions;
+
+use DOMDocument;
+use SAML2\Utils;
+
+class GsspUserAttributesChunk extends Chunk {
+
+    public function __construct(\DOMElement $value = null)
+    {
+        if ($value === null) {
+            $doc = new DOMDocument("1.0", "UTF-8");
+            $root = $doc->createElementNS('urn:mace:surf.nl:stepup:gssp-extensions', 'gssp:UserAttributes');
+            $doc->appendChild($doc->importNode($root));
+            $value = $doc->documentElement;
+        }
+        parent::__construct('UserAttributes', 'urn:mace:surf.nl:stepup:gssp-extensions', $value);
+    }
```

Comment:
Comparing your XML output with that proposed in the RFC I see some small differences.

1. The `UserAttributes` element is not registered with the `xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"` and `xmlns:xs="http://www.w3.org/2001/XMLSchema"` namespaces.
2. The `Attribute` element is registered with the `urn:oasis:names:tc:SAML:2.0:assertion` namespace. This should not be necessary.
3. The `AttributeValue` element only requires the data type to be set as an attribute. Additional namespace registrations where set on the attribute value attribute.

Reply:
> Comparing your XML output with that proposed in the RFC I see some small differences.
> 
> 1. The `UserAttributes` element is not registered with the `xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"` and `xmlns:xs="http://www.w3.org/2001/XMLSchema"` namespaces.
> 2. The `Attribute` element is registered with the `urn:oasis:names:tc:SAML:2.0:assertion` namespace. This should not be necessary.
> 3. The `AttributeValue` element only requires the data type to be set as an attribute. Additional namespace registrations where set on the attribute value attribute.

Yes, this only adds the namespace on the attributes that actually use it. That is due to the way its being rendered by the domdocument and being able to just add one attribute at the time. As far as know it does not matter and since the parsing by the simplesaml library is ok i assume it is no issue.

Reply:
Yeah, should be good to go, if this poses problems in the course of the integration we can always adjust at that point.